// @wellwind/ngx-thanos/esm5/lib/thanos-snap.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, ElementRef, EventEmitter, Input, Output } from '@angular/core';
import { ThanosSnapService } from './thanos-snap.service';
var ThanosSnapDirective = /** @class */ (function () {
    function ThanosSnapDirective(elementRef, thanosSnapService) {
        this.elementRef = elementRef;
        this.thanosSnapService = thanosSnapService;
        this.ngxThanosSnap = false;
        this.snapStarting = new EventEmitter();
        this.snapEnded = new EventEmitter();
        this.rewinded = new EventEmitter();
    }
    Object.defineProperty(ThanosSnapDirective.prototype, "host", {
        get: /**
         * @return {?}
         */
        function () {
            return this.elementRef.nativeElement;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    ThanosSnapDirective.prototype.ngOnChanges = /**
     * @return {?}
     */
    function () {
        if (this.ngxThanosSnap) {
            this.snap();
        }
        else {
            this.rewind();
        }
    };
    /**
     * @return {?}
     */
    ThanosSnapDirective.prototype.snap = /**
     * @return {?}
     */
    function () {
        var _this = this;
        this.snapStarting.emit();
        this.thanosSnapService.snap(this.host).subscribe((/**
         * @return {?}
         */
        function () { return _this.snapEnded.emit(); }));
    };
    /**
     * @return {?}
     */
    ThanosSnapDirective.prototype.rewind = /**
     * @return {?}
     */
    function () {
        var _this = this;
        this.thanosSnapService.rewind(this.host).subscribe((/**
         * @return {?}
         */
        function () { return _this.rewinded.emit(); }));
    };
    ThanosSnapDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[ngxThanosSnap]',
                    exportAs: 'ngxThanosSnap'
                },] }
    ];
    /** @nocollapse */
    ThanosSnapDirective.ctorParameters = function () { return [
        { type: ElementRef },
        { type: ThanosSnapService }
    ]; };
    ThanosSnapDirective.propDecorators = {
        ngxThanosSnap: [{ type: Input }],
        snapStarting: [{ type: Output }],
        snapEnded: [{ type: Output }],
        rewinded: [{ type: Output }]
    };
    return ThanosSnapDirective;
}());
export { ThanosSnapDirective };
if (false) {
    /** @type {?} */
    ThanosSnapDirective.prototype.ngxThanosSnap;
    /** @type {?} */
    ThanosSnapDirective.prototype.snapStarting;
    /** @type {?} */
    ThanosSnapDirective.prototype.snapEnded;
    /** @type {?} */
    ThanosSnapDirective.prototype.rewinded;
    /**
     * @type {?}
     * @private
     */
    ThanosSnapDirective.prototype.elementRef;
    /**
     * @type {?}
     * @private
     */
    ThanosSnapDirective.prototype.thanosSnapService;
}
