// @wellwind/ngx-thanos/esm2015/lib/thanos-snap.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, ElementRef, EventEmitter, Input, Output } from '@angular/core';
import { ThanosSnapService } from './thanos-snap.service';
export class ThanosSnapDirective {
    /**
     * @param {?} elementRef
     * @param {?} thanosSnapService
     */
    constructor(elementRef, thanosSnapService) {
        this.elementRef = elementRef;
        this.thanosSnapService = thanosSnapService;
        this.ngxThanosSnap = false;
        this.snapStarting = new EventEmitter();
        this.snapEnded = new EventEmitter();
        this.rewinded = new EventEmitter();
    }
    /**
     * @return {?}
     */
    get host() {
        return this.elementRef.nativeElement;
    }
    /**
     * @return {?}
     */
    ngOnChanges() {
        if (this.ngxThanosSnap) {
            this.snap();
        }
        else {
            this.rewind();
        }
    }
    /**
     * @return {?}
     */
    snap() {
        this.snapStarting.emit();
        this.thanosSnapService.snap(this.host).subscribe((/**
         * @return {?}
         */
        () => this.snapEnded.emit()));
    }
    /**
     * @return {?}
     */
    rewind() {
        this.thanosSnapService.rewind(this.host).subscribe((/**
         * @return {?}
         */
        () => this.rewinded.emit()));
    }
}
ThanosSnapDirective.decorators = [
    { type: Directive, args: [{
                selector: '[ngxThanosSnap]',
                exportAs: 'ngxThanosSnap'
            },] }
];
/** @nocollapse */
ThanosSnapDirective.ctorParameters = () => [
    { type: ElementRef },
    { type: ThanosSnapService }
];
ThanosSnapDirective.propDecorators = {
    ngxThanosSnap: [{ type: Input }],
    snapStarting: [{ type: Output }],
    snapEnded: [{ type: Output }],
    rewinded: [{ type: Output }]
};
if (false) {
    /** @type {?} */
    ThanosSnapDirective.prototype.ngxThanosSnap;
    /** @type {?} */
    ThanosSnapDirective.prototype.snapStarting;
    /** @type {?} */
    ThanosSnapDirective.prototype.snapEnded;
    /** @type {?} */
    ThanosSnapDirective.prototype.rewinded;
    /**
     * @type {?}
     * @private
     */
    ThanosSnapDirective.prototype.elementRef;
    /**
     * @type {?}
     * @private
     */
    ThanosSnapDirective.prototype.thanosSnapService;
}
